src/oracle/connection/stmt.rs

```diff
@@ -1,209 +1,585 @@
+
+            println!("{:?}", mysql);
+        }
+    }
+    mysql
 }
 
 impl Statement {
     pub fn prepare(raw_connection: &Rc<RawConnection>, sql: &str) -> QueryResult<Self> {
-        println!("prepare statment {}", sql);
+        //let mysql = reform_insert_query2(sql);
+        let mysql = sql.to_string();
+
+        println!("prepare statement {}", mysql);
         let stmt = unsafe {
             let mut stmt: *mut ffi::OCIStmt = ptr::null_mut();
-            ffi::OCIStmtPrepare2(raw_connection.service_handle,
-                                 &mut stmt,
-                                 raw_connection.env.error_handle,
-                                 sql.as_ptr(),
-                                 sql.len() as u32,
-                                 ptr::null(),
-                                 0,
-                                 ffi::OCI_NTV_SYNTAX,
-                                 ffi::OCI_DEFAULT);
+            let status = ffi::OCIStmtPrepare2(
+                raw_connection.service_handle,
+                &mut stmt,
+                raw_connection.env.error_handle,
+                mysql.as_ptr(),
+                mysql.len() as u32,
+                ptr::null(),
+                0,
+                ffi::OCI_NTV_SYNTAX,
+                ffi::OCI_DEFAULT,
+            );
+
+            if let Some(err) = Self::check_error(raw_connection.env.error_handle, status) {
+                return Err(err);
+            }
+
+            // for create statements we need to run OCIStmtPrepare2 twice
+            // c.f. https://docs.oracle.com/database/121/LNOCI/oci17msc001.htm#LNOCI17165
+            // "To reexecute a DDL statement, you must prepare the statement again using OCIStmtPrepare2()."
+            if let Some(u) = mysql.to_string().find("CREATE") {
+                if u < 10 {
+                    let status = ffi::OCIStmtPrepare2(
+                        raw_connection.service_handle,
+                        &mut stmt,
+                        raw_connection.env.error_handle,
+                        mysql.as_ptr(),
+                        mysql.len() as u32,
+                        ptr::null(),
+                        0,
+                        ffi::OCI_NTV_SYNTAX,
+                        ffi::OCI_DEFAULT,
+                    );
+
+                    if let Some(err) = Self::check_error(raw_connection.env.error_handle, status) {
+                        return Err(err);
+                    }
+                }
+            }
+
             stmt
         };
         Ok(Statement {
-               connection: raw_connection.clone(),
-               inner_statement: stmt,
-               bind_index: 0,
-               is_select: sql.contains("SELECT"),
-           })
+            connection: raw_connection.clone(),
+            inner_statement: stmt,
+            bind_index: 0,
+            is_select: sql.contains("SELECT") || sql.contains("select"),
+            binds: Vec::with_capacity(20),
```

Comment:
In general I can't imagine this without any SQL string parsing which we want to avoid here. Since we later use `push` I don't see a problem here in allocating 20 elements in the first place. I moved this into a `const` and maybe we use 1024 or something more meaningful here.